pkg/controllers/subnetport/subnetport_controller.go

```diff
@@ -0,0 +1,290 @@
+/* Copyright © 2022 VMware, Inc. All Rights Reserved.
+   SPDX-License-Identifier: Apache-2.0 */
+
+package subnetport
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"reflect"
+	"runtime"
+
+	"github.com/vmware-tanzu/nsx-operator/pkg/logger"
+	"github.com/vmware-tanzu/nsx-operator/pkg/metrics"
+	v1 "k8s.io/api/core/v1"
+	apimachineryruntime "k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/util/workqueue"
+	ctrl "sigs.k8s.io/controller-runtime"
+
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	"github.com/vmware-tanzu/nsx-operator/pkg/apis/v1alpha1"
+	"github.com/vmware-tanzu/nsx-operator/pkg/controllers/common"
+	servicecommon "github.com/vmware-tanzu/nsx-operator/pkg/nsx/services/common"
+	"github.com/vmware-tanzu/nsx-operator/pkg/nsx/services/subnetport"
+)
+
+var (
+	log                     = logger.Log
+	MetricResTypeSubnetPort = common.MetricResTypeSubnetPort
+)
+
+// SubnetPortReconciler reconciles a SubnetPort object
+type SubnetPortReconciler struct {
+	client.Client
+	Scheme  *apimachineryruntime.Scheme
+	Service *subnetport.SubnetPortService
+}
+
+//+kubebuilder:rbac:groups=nsx.vmware.com,resources=subnetports,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=nsx.vmware.com,resources=subnetports/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=nsx.vmware.com,resources=subnetports/finalizers,verbs=update
+func (r *SubnetPortReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	obj := &v1alpha1.SubnetPort{}
+	log.Info("reconciling subnetport CR", "subnetport", req.NamespacedName, "obj", obj)
+
+	metrics.CounterInc(r.Service.NSXConfig, metrics.ControllerSyncTotal, MetricResTypeSubnetPort)
+
+	if err := r.Client.Get(ctx, req.NamespacedName, obj); err != nil {
+		log.Error(err, "unable to fetch subnetport CR", "req", req.NamespacedName)
+		return common.ResultNormal, client.IgnoreNotFound(err)
+	}
+
+	if obj.ObjectMeta.DeletionTimestamp.IsZero() {
+		metrics.CounterInc(r.Service.NSXConfig, metrics.ControllerUpdateTotal, MetricResTypeSubnetPort)
+		if !controllerutil.ContainsFinalizer(obj, servicecommon.FinalizerName) {
+			controllerutil.AddFinalizer(obj, servicecommon.FinalizerName)
+			if err := r.Client.Update(ctx, obj); err != nil {
+				log.Error(err, "add finalizer", "subnetport", req.NamespacedName)
+				updateFail(r, &ctx, obj, &err)
+				return common.ResultRequeue, err
+			}
+			log.V(1).Info("added finalizer on subnetport CR", "subnetport", req.NamespacedName)
+		}
+
+		if len(obj.Spec.SubnetSet) > 0 && len(obj.Spec.Subnet) > 0 {
```

Comment:
In current subnet CR implementation, it's possible that both `.spec.subnetset` and `.spec.subnet` are specified. In this scenario, `.spec.subnetset` are specified by user and `subnetport_handler` will fill field `.spec.subnet` after an available subnet under the subnetset is allocated. So there  will be conflicts here. Maybe `subnetport_handler` should use another field like `annotation` to mark the subnet allocated from subnetset. any suggestion?

https://github.com/vmware-tanzu/nsx-operator/pull/165/files#diff-959aec2764d2baff7b43c068f97bee2f29c1c0c7e8bc8554569346cdf32499ccR48

Reply:
In the latest design there won't be `subnetport_handler` filling `.spec.subnet` so mark it as resolved.